warehouse/transformer/benchmark/main.go

```diff
@@ -0,0 +1,143 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/google/uuid"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	"github.com/rudderlabs/rudder-go-kit/profiler"
+	"github.com/rudderlabs/rudder-go-kit/stats"
+
+	"github.com/rudderlabs/rudder-server/jsonrs"
+	ptrans "github.com/rudderlabs/rudder-server/processor/transformer"
+	"github.com/rudderlabs/rudder-server/processor/types"
+	"github.com/rudderlabs/rudder-server/utils/timeutil"
+	wtrans "github.com/rudderlabs/rudder-server/warehouse/transformer"
+)
+
+const (
+	processorTransformer = "PROCESSOR_TRANSFORMER"
+	warehouseTransformer = "WAREHOUSE_TRANSFORMER"
+)
+
+func main() {
+	ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGKILL, syscall.SIGINT, syscall.SIGTERM)
+	defer cancel()
+
+	if err := run(ctx); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func run(ctx context.Context) error {
+	sampleEvent := mustString("SAMPLE_EVENT")
+	eventsInBatch, batchSize := mustInt("EVENTS_IN_BATCH"), mustInt("BATCH_SIZE")
+	iterations := mustInt("NUM_OF_ITERATIONS")
+	mode := mustString("MODE")
+
+	if eventsInBatch < batchSize {
+		return errors.New("eventsInBatch must be greater than or equal to batchSize")
+	}
+
+	runID := uuid.New().String()
+	conf := config.Default
+
+	logFactory := logger.Default
+	l := logFactory.NewLogger().Child("warehouse-transformer-benchmark").Child(mode)
+
+	clientEvents := make([]types.TransformerEvent, eventsInBatch)
+	for i := 0; i < len(clientEvents); i++ {
+		var transformerEvent types.TransformerEvent
+		err := jsonrs.Unmarshal([]byte(sampleEvent), &transformerEvent)
+		if err != nil {
+			return fmt.Errorf("could not unmarshal sample client event: %w", err)
+		}
+		clientEvents[i] = transformerEvent
+	}
+
+	t, err := selectTransformer(mode, conf, l)
+	if err != nil {
+		return fmt.Errorf("could not select transformer: %w", err)
+	}
+
+	startTime := timeutil.Now()
+
+	l.Infon("Starting transformer")
+	l.Infof("Started At: %s", startTime.Format(time.RFC3339))
+	l.Infof("Sample Event: %s", sampleEvent)
+	l.Infof("Events in Batch: %d", eventsInBatch)
+	l.Infof("Batch size: %d", batchSize)
+	l.Infof("Iterations: %d", iterations)
+	l.Infof("Mode: %s", mode)
+	l.Infof("RunID: %s", runID)
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	g, gCtx := errgroup.WithContext(ctx)
+	g.Go(func() error {
+		return profiler.StartServer(gCtx, config.GetInt("Profiler.Port", 7777))
+	})
+	g.Go(func() error {
+		for i := 0; i < iterations; i++ {
+			l.Infof("Running iteration: %d", i+1)
+			t.Transform(gCtx, clientEvents, batchSize)
+		}
+
+		endTime := timeutil.Now().Sub(startTime).Seconds()
+
+		l.Infof("Total events transformer: %d", len(clientEvents)*iterations)
+		l.Infof("Total time for transformations: %v", endTime)
+		l.Infof("Transformation rate (msg/s): %v",
```

Comment:
How are we using this benchmark to make decisions? Are we evaluating the `transformation rate` for each input combination?

Reply:
Yeah, we are evaluating the **Transformation rate** along with the **Kubernetes Compute Resources** (CPU and Memory) and the **Pyroscope Profiles** (CPU and Memory).